x/cron/cron.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	for {
+		key := queueKey(runAt)
+		if ok, err := db.Has(key); err != nil {
+			return nil, errors.Wrap(err, "cannot check key existance")
+		} else if ok {
+			// If the key is already in use, instead of storing a
+			// list of messages under each key, which is a very
+			// unlikely to happen, increase the execution time by
+			// the smallest duration.
+			// Message is guaranteed to be executed not earlier
+			// than given time, NOT at exactly given time.
+			runAt = runAt.Add(granularity)
+			continue
+		}
+
+		if err := db.Set(key, rawTask); err != nil {
+			return nil, errors.Wrap(err, "cannot store in queue")
+		}
+		return key, nil
+	}
+}
+
+func queueKey(t time.Time) []byte {
+	rawTime := make([]byte, 8)
+	// Zero time does not need to put any data as the bytes are already set
+	// to zero.
+	if !t.IsZero() {
+		binary.BigEndian.PutUint64(rawTime, uint64(t.UnixNano()))
+	}
+	return append([]byte("_crontx:runat:"), rawTime...)
+}
+
+// Ticker allows to execute messages queued for future execution. It does this
+// by implementing weave.Ticker interface.
+type Ticker struct {
+	hn       weave.Handler
+	taskType reflect.Type
+	results  orm.ModelBucket
+}
+
+// NewTicker returns a cron instance that is using given handler to process all
+// queued messages that execution time is due.
+//
+// Task is an instance of a Task interface implementation. This is the
+// implementation that the returned ticker can work with and no other
+// implementation is allowed. All data serialization and processing will be
+// done assuming provided type.
+func NewTicker(emptyInstance Task, h weave.Handler) *Ticker {
+	return &Ticker{
+		hn:       h,
+		taskType: reflect.TypeOf(emptyInstance),
+		results:  NewTaskResultBucket(),
+	}
+}
+
+var _ weave.Ticker = (*Ticker)(nil)
+
+// Tick implementes weave.Ticker interface.
+// Tick can process any number of messages suitable for execution. All changes
+// are done atomically and apply only on success.
+func (t *Ticker) Tick(ctx context.Context, db store.CacheableKVStore) [][]byte {
+	executed, err := t.tick(ctx, db)
+	if err != nil {
+		// TODO log error?
+	}
+	return executed
+}
+
+// tick process any number of tasks. It always returns a response and might
+// return an error. This method is similar to the Tick except it provides an
+// error. This makes it easier for the tests to check the result.
+func (t *Ticker) tick(ctx context.Context, db store.CacheableKVStore) ([][]byte, error) {
+	var executed [][]byte
+
+	now, err := weave.BlockTime(ctx)
+	if err != nil {
+		return executed, errors.Wrap(err, "cannot get current time")
+	}
+
+	for {
+		task := reflect.New(t.taskType.Elem()).Interface().(Task)
+		switch key, err := peek(db, now, task); {
+		case err == nil:
+			res := TaskResult{
+				Metadata:   &weave.Metadata{Schema: 1},
+				Successful: true,
+			}
+
+			taskCtx := ctx
+			if prep, ok := task.(ContextPreparer); ok {
+				taskCtx = prep.Prepare(taskCtx)
```

Comment:
@ethanfrey I am trying to figure out a good way to insert authentication (and other) information into the context. My latest approach is to delegate this to the client.

I have realized that this cannot be delegated to the client completely, because it is heavily dependent on the authentication used by the router. To keep things simple I was planning to use [the same router as the rest of the application](https://github.com/iov-one/weave/pull/837/files#diff-b5f4d55718db6c007d81b069f384af00R139). This means, that the same `sigs` and `multisig` authentication is required.

I wonder if it would not be best to use the same technique for the [`Task` implementation](https://github.com/iov-one/weave/pull/837/files#diff-30e5ae30dd16577b041287ae8c5b734cR144) - create a message with middleware fields like `multisig` and `signatures` and use extension decorators to work with them. This make the whole functionality very much like a synchronous request handling, which is good.
The problem with this approach is that extension cannot set [`Task` attributes](https://github.com/iov-one/weave/pull/837/files#diff-30e5ae30dd16577b041287ae8c5b734cR144) because it does not know about the implementation. :thinking: